kernel/src/actions/mod.rs

```diff
@@ -423,26 +423,39 @@ impl Protocol {
 
     /// Get the reader features for the protocol
     #[internal_api]
-    pub(crate) fn reader_features(&self) -> Option<&[ReaderFeature]> {
+    pub(crate) fn reader_features(&self) -> Option<&[TableFeature]> {
         self.reader_features.as_deref()
     }
 
     /// Get the writer features for the protocol
     #[internal_api]
-    pub(crate) fn writer_features(&self) -> Option<&[WriterFeature]> {
+    pub(crate) fn writer_features(&self) -> Option<&[TableFeature]> {
         self.writer_features.as_deref()
     }
 
     /// True if this protocol has the requested reader feature
-    pub(crate) fn has_reader_feature(&self, feature: &ReaderFeature) -> bool {
+    /// Note: Use `has_table_feature` instead if given feature is supported for both read/write.
+    pub(crate) fn has_reader_feature(&self, feature: &TableFeature) -> bool {
         self.reader_features()
             .is_some_and(|features| features.contains(feature))
     }
 
-    /// True if this protocol has the requested writer feature
-    pub(crate) fn has_writer_feature(&self, feature: &WriterFeature) -> bool {
-        self.writer_features()
-            .is_some_and(|features| features.contains(feature))
+    /// True if this protocol has the requested table feature
+    /// - ReaderWriter features must be in both reader features and writer features.
+    /// - Writer features only need to be in writer features.
+    pub(crate) fn has_table_feature(&self, feature: &TableFeature) -> bool {
+        match feature.feature_type() {
```

Comment:
Can you add a check to the protocol that ensures that:
* check that all reader features are reader/writer. reader_features.all(|feature| matches!(feature.get_type(), FeatureType::ReaderWriter))
* Check that all writer features are writer. writer_feature.all(|feature| matches!(feature.get_type(), FeatureType::Writer))
* Check that the set of reader features is a subset of writer  features

Then this check just needs to check presence in the write features.

Reply:
Thank you for review. I made the changes based on your suggestions and https://github.com/delta-io/delta-kernel-rs/pull/1290#discussion_r2373665696